osgar/lib/route.py

```diff
@@ -0,0 +1,151 @@
+"""
+  Route with geo2plan and plan2geo conversions + snap position.
+"""
+
+import math
+from osgar.lib.line import Line, distance, pointAtPolyLineDist
+
+
+EARTHRADIUS = 6.380935276e+06
+SUIDARHTRAE = 1.567168380e-07
+DEGRAD            = 1.7453292519943295769236907684886e-02
+RADDEG            = 5.7295779513082320876798154814105e+01
+
+
+def loadLatLonPtsFromStr(input_str):
+    lines = input_str.split("\n")
+    pts = []
+    for line in lines:
+        if len(line) > 0 and line[0] != '#':
+            assert( len(line.split()) == 2 )
+            lat, lon = line.split()
+            pts.append( (float(lon), float(lat)) )
+    return pts
+
+
+def loadLatLonPts(filename):
+    with open(filename, "r") as f:
+        return loadLatLonPtsFromStr(f.read())
+
+
+def saveLatLonPts(pts, filename, note = None):
+    with open(filename, "w") as f:
+        if note:
+            f.write("# " + note + "\n")
+        for p in pts:
+            f.write("%f %f\n" % (p[1], p[0]))
+
+
+class Convertor:
+    "convert lat/lon and planar coordinates"
+    def __init__(self, refPoint = (16.60906, 49.2060633333)):
+        self._cosMulti = max(0.001, EARTHRADIUS*math.cos(refPoint[1]*DEGRAD));
+        self.refPoint = refPoint
+
+    def geo2planar(self, pos):
+        return ((pos[0] - self.refPoint[0]) * DEGRAD * self._cosMulti, (pos[1] - self.refPoint[1]) * DEGRAD * EARTHRADIUS)
+
+    def planar2geo(self, pos):
+        return (self.refPoint[0] + pos[0] * RADDEG/ self._cosMulti, self.refPoint[1] + pos[1] * RADDEG * SUIDARHTRAE)
+
+
+class DummyConvertor:
+    "convert 1:1 for coordinates already in meters"
+    def geo2planar(self, pos):
+        return pos
+
+    def planar2geo(self, pos):
+        return pos
+
+
+class Route:
+    "defined route for navigation"
+    def __init__(self, pts = [], conv = None, isLoop = None):
+        if conv == None:
+            conv = Convertor()
+        self.conv = conv
+        self.pts = [ self.conv.geo2planar(p) for p in pts ]
+        if isLoop == None and len(self.pts) > 1:
+            isLoop = (distance(self.pts[0], self.pts[-1]) < 5.0)
+        self.isLoop = isLoop
+
+    def length( self ):
+        if not self.pts:
+            return None
+        sum = 0
+        prev = self.pts[0]
+        for p in self.pts:
+            sum += distance( prev, p )
+            prev = p
+        return sum
+
+    def findNearestEx( self, pos ):
```

Comment:
Can you please add a comment about the purpose of the function? I am not really sure from the code.